// This file is made available under Elastic License 2.0
// This file is based on code available under the Apache license here:
//   https://github.com/apache/incubator-doris/blob/master/gensrc/proto/olap_file.proto

// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// Define file format struct, like data header, index header.

syntax="proto2";

package starrocks;
option java_package = "com.starrocks.proto";

import "olap_common.proto";
import "types.proto";

message ZoneMapOfRowset {
    required bytes min = 1;
    required bytes max = 2;
    optional bool null_flag = 3;
}

enum RowsetTypePB {
    ALPHA_ROWSET = 0; // Deleted
    BETA_ROWSET  = 1;
}

enum RowsetStatePB {
    // Rowset be writting
    PREPARED = 0;
    // indicates that the rowset is written, but not yet visible to the user; the rowset in this
    // state cannot be deleted by the BE itself, but must be commanded by the FE
    COMMITTED = 1;
    // Indicates that the rowset is already visible to the user
    VISIBLE = 2;
}

// indicate whether date between segments of a rowset is overlappinng
enum SegmentsOverlapPB {
    OVERLAP_UNKNOWN = 0;
    OVERLAPPING = 1;
    NONOVERLAPPING = 2;
}

// indicate segment footer position and size of partial rowset
message FooterPointerPB {
    optional uint64 position = 1;
    optional uint64 size = 2;
}

// some txn semantic information bind to a written rowset
// i.e.
//   1. partial update txn only provide and update a subset of columns,
//      so it should add info about updated columns, whether to insert new
//      rows if old row doesn't exists, and default values etc
//   2. conditional update txn may add some meta to describe condition
//   3. more general read-write txn may add complex expressions
message RowsetTxnMetaPB {
    repeated uint32 partial_update_column_ids = 1;
    repeated uint32 partial_update_column_unique_ids = 2;

    // one for each segment
    repeated FooterPointerPB partial_rowset_footers = 3;
}

message RowsetMetaPB {
    required int64 rowset_id = 1;
    optional int64 partition_id = 2;
    optional int64 tablet_id = 3;
    // only for pending rowset
    optional int64 txn_id = 4;
    // tablet id and schema hash could find tablet
    optional int32 tablet_schema_hash = 5;
    optional RowsetTypePB rowset_type = 6;
    optional RowsetStatePB rowset_state = 7;
    // only for visible rowset.
    optional int64 start_version = 8;
    // only for visible rowset.
    optional int64 end_version = 9;
    // only for visible rowset.
    optional int64 version_hash = 10 [deprecated = true];
    // total number of rows.
    optional int64 num_rows = 11;
    // calculated by index + data
    optional int64 total_disk_size = 12;
    // calculated sum(segmentgroup.data_size)
    optional int64 data_disk_size = 13;
    // calculated sum(segmentgroup.index_size)
    optional int64 index_disk_size = 14;
    // rowset level column min/max/null statistics, not not used currently
    repeated ZoneMapOfRowset zone_maps = 15 [deprecated = true];
    optional DeletePredicatePB delete_predicate = 16;
    // calculated from segment group
    optional bool empty = 17;
    // only for pending rowset
    optional PUniqueId load_id = 18;
    // not set, get from olap index header msg?
    optional bool delete_flag = 19;
    optional int64 creation_time = 20;
    optional PUniqueId tablet_uid = 21;
    // total number of segments
    optional int64 num_segments = 22;
    // rowset id definition, it will replace required rowset id
    optional string rowset_id_v2 = 23;
    // to indicate whether the data between the segments overlap
    optional SegmentsOverlapPB segments_overlap_pb = 51 [default = OVERLAP_UNKNOWN];
    // every segment in a rowset has an unique uint32 id
    // it is incrementally allocated when the rowset is committed
    // if a rowset have 4 segments, 4 ids will be allocated.
    // if rowset_seg_id=n then segment ids will be: n, n+1, n+2, n+3
    optional uint32 rowset_seg_id = 52;
    // number of delete files, delete file store delete operations'
    // primary keys
    optional uint32 num_delete_files = 53;
    // total row size in approximately
    optional int64 total_row_size = 54;
    // some txn semantic information bind to this rowset
    optional RowsetTxnMetaPB txn_meta = 55;
}

enum DataFileType {
    OLAP_DATA_FILE = 0; //Deprecated. Only columnar-wise format is supported.
    COLUMN_ORIENTED_FILE = 1;
}

// Please keep the defined value less than 256, bacause this enum will be casted to uint8_t
// in some cases.
enum KeysType {
    DUP_KEYS = 0;
    UNIQUE_KEYS = 1;
    AGG_KEYS = 2;
    PRIMARY_KEYS = 10;
}

message DeletePredicatePB {
    required int32 version = 1;
    repeated string sub_predicates = 2;
    repeated InPredicatePB in_predicates = 3;
}

message InPredicatePB {
    optional string column_name = 1;
    optional bool is_not_in = 2;
    repeated string values = 3;
}

enum AlterTabletState {
    ALTER_PREPARED = 0;
    ALTER_RUNNING = 1;
    ALTER_FINISHED = 2;
    ALTER_FAILED = 3;
}

enum AlterTabletType {
    SCHEMA_CHANGE = 1;
    ROLLUP = 2;
}

message AlterTabletPB {
    optional AlterTabletState alter_state = 1;
    required int64 related_tablet_id = 2;
    optional int32 related_schema_hash = 3;
    optional AlterTabletType alter_type = 4;
}

message ColumnPB {
    required int32 unique_id = 1; // ColumnMessage.unique_id
    optional string name = 2; // ColumnMessage.name
    required string type = 3; // ColumnMessage.type
    optional bool is_key = 4; // ColumnMessage.is_key
    optional string aggregation = 5; // ColumnMessage.aggregation
    optional bool is_nullable = 6; // ColumnMessage.is_allow_null
    optional bytes default_value = 7; // ColumnMessage.default_value ?
    optional int32 precision = 8; // ColumnMessage.precision
    optional int32 frac = 9; // ColumnMessage.frac
    optional int32 length = 10; // ColumnMessage.length
    optional int32 index_length = 11; // ColumnMessage.index_length
    optional bool is_bf_column = 12; // ColumnMessage.is_bf_column
    optional int32 referenced_column_id = 13; //
    optional string referenced_column = 14; // ColumnMessage.referenced_column?
    optional bool has_bitmap_index = 15 [default=false]; // ColumnMessage.has_bitmap_index
    optional bool visible = 16 [default=true]; // used for hided column
    repeated ColumnPB children_columns = 17;
}

message TabletSchemaPB {
    optional KeysType keys_type = 1;    // OLAPHeaderMessage.keys_type
    repeated ColumnPB column = 2;   // OLAPHeaderMessage.column
    optional int32 num_short_key_columns = 3;   // OLAPHeaderMessage.num_short_key_fields
    optional int32 num_rows_per_row_block = 4;  // OLAPHeaderMessage.num_rows_per_data_block
    optional CompressKind compress_kind = 5; // OLAPHeaderMessage.compress_kind
    optional double bf_fpp = 6; // OLAPHeaderMessage.bf_fpp
    optional uint32 next_column_unique_id = 7; // OLAPHeaderMessage.next_column_unique_id
    optional bool DEPRECATED_is_in_memory = 8 [default=false];
    optional int64 deprecated_id = 9; // deprecated
    optional int64 id = 50;
}

enum TabletStatePB {
    PB_NOTREADY  = 0; // under alter table, rollup, clone
    PB_RUNNING = 1;
    PB_TOMBSTONED = 2;
    PB_STOPPED = 3;
    PB_SHUTDOWN = 4;
}

enum TabletTypePB {
    TABLET_TYPE_DISK = 0;
    TABLET_TYPE_MEMORY = 1;
}

message CompactionInfoPB {
    optional EditVersionPB start_version = 1;
    repeated uint32 inputs = 2;
    // may support multiple output rowsets in future
    repeated uint32 outputs = 3;
}

message EditVersionMetaPB {
    optional EditVersionPB version = 1;
    optional int64 creation_time = 2;
    repeated uint32 rowsets = 3; // full data, optional
    repeated uint32 rowsets_add = 4; // incremental optional
    repeated uint32 rowsets_del = 5; // incremental optional
    repeated uint32 deltas = 6;
    optional CompactionInfoPB compaction = 7;
    optional uint32 rowsetid_add = 8; // only available&useful in meta log
}

enum TabletMetaOpType {
    OP_NONE = 0;
    OP_ROWSET_COMMIT = 1;
    OP_COMPACTION_COMMIT = 2;
    OP_APPLY = 3;
}

message TabletMetaOpPB {
    optional TabletMetaOpType type = 1;
    optional EditVersionMetaPB commit = 2;
    optional EditVersionPB apply = 3;
}

message TabletMetaLogPB {
    repeated TabletMetaOpPB ops = 1;
}

message TabletUpdatesPB {
    repeated EditVersionMetaPB versions = 1;
    optional EditVersionPB apply_version = 2;
    optional uint32 next_rowset_id = 3;
    optional uint64 next_log_id = 4;
}

message TabletMetaPB {
    optional int64 table_id = 1;    // ?
    optional int64 partition_id = 2;    // ?
    optional int64 tablet_id = 3;   // OlapHeaderMessage.tablet_id
    optional int32 schema_hash = 4; // OlapHeaderMessage.schema_hash
    optional int32 shard_id = 5;    // OlapHeaderMessage.shard
    optional int64 creation_time = 6;   // OlapHeaderMessage.creation_time
    optional int64 cumulative_layer_point = 7;  // OlapHeaderMessage.cumulative_layer_point

    optional TabletStatePB tablet_state = 8;
    optional TabletSchemaPB schema = 9;
    repeated RowsetMetaPB rs_metas = 10;
    repeated RowsetMetaPB inc_rs_metas = 11;
    optional AlterTabletPB alter_task = 12;
    // if true, this tablet will not do compaction,
    // and does not create init version
    optional bool in_restore_mode = 13 [default = false];   // OlapHeaderMessage.in_restore_mode
    // a uniqued id to identified tablet with same tablet_id and schema hash
    optional PUniqueId tablet_uid = 14;
    optional int64 end_rowset_id = 15;
    optional RowsetTypePB preferred_rowset_type = 16;
    optional TabletTypePB tablet_type = 17;
    optional TabletUpdatesPB updates = 50; // used for new updatable tablet
}

message OLAPIndexHeaderMessage {
    required int32 start_version = 1;
    required int32 end_version = 2;
    required int64 cumulative_version_hash = 3 [deprecated = true];

    required uint32 segment = 4;
    required uint32 num_rows_per_block = 5;
    optional bool null_supported = 6;
    optional bool delete_flag = 7;
}

message OLAPDataHeaderMessage {
    required uint32 segment = 2;
}

message OLAPRawDeltaHeaderMessage {
    required int32 schema_hash = 2;
}
